### PR TITLE
[GDNative] (version-) error reporting + small API change

### LINKED ISSUE #12650: Dangling pointer bug in `godot_string_cstr`
https://github.com/godotengine/godot/blob/a89fa34c21103430b1d140ee04c3ae6a433d77ce/modules/gdnative/gdnative/string.cpp#L95-L98

What happens here is this call chain:

- `godot_string_cstr`
- [`auto String::utf8() const -> CharString`](https://github.com/godotengine/godot/blob/a89fa34c21103430b1d140ee04c3ae6a433d77ce/core/ustring.cpp#L1409-L1488)
- [`auto CharString::get_data() const -> *const char`](https://github.com/godotengine/godot/blob/a89fa34c21103430b1d140ee04c3ae6a433d77ce/core/ustring.cpp#L89-L95)
- `return`
- `CharString::~CharString()`
- [`Vector<char>::~Vector()`](https://github.com/godotengine/godot/blob/a89fa34c21103430b1d140ee04c3ae6a433d77ce/core/vector.h#L422-L426)
- [`void Vector<char>::_unref(void*)`](https://github.com/godotengine/godot/blob/master/core/vector.h#L188-L210) <== Memory deleted here, because `CharString` went out of scope and only had a ref count of 1.

So, `String` parses its UTF-32 data into a UTF-8 buffer, returns a pointer to the UTF-8 buffer's data, and as the function `godot_string_cstr` returns, that buffer is deleted, resulting in a dangling pointer.

I currently see only four solutions to that bug, one being less elegant than the previous one:

- Cache the created `CharString`, which keeps the reference count above 1. The problem with this is that the created UTF-8 data will be kept until the corresponding `String` dies, increasing memory consumption by a bit. Shouldn't be a huge problem, though, unless someone `godot_string_cstr`s mebibytes of text all the time. After all, UTF-8 data is on average only 1/4th to 1/2th the size of UTF-32 data.
- `memcpy` the buffer into one that the user has to de-allocate himself with a call to `godot_free`. `memcpy` cannot be avoided here, instead of just stealing a `Vector<T>`'s `_ptr`, as `godot_free` doesn't have the boolean flag for de-allocating ref counted buffers.
- Remove that function entirely *(We don't want that to happen, do we?)*
- Expose `CharString` to GDNative *(I don't think that we want that, either)*

### PR DESCRIPTION
**Pull request by karroffel**
fixes #12650

### CODE REVIEW

**Comment by karroffel** [comment]
cc @endragor you might want to see if that error checking is sufficient.

File: modules/gdnative/gdnative/gdnative.cpp [hunk 1]
```diff
@@ -137,6 +139,32 @@ void GDAPI godot_print(const godot_string *p_message) {
...
+	message += String("Got version {have_major}.{have_minor} but needs {want_major}.{want_minor}!").format(versions);
+
+	_err_print_error("gdnative_init", library->get_current_library_path().utf8().ptr(), 0, message.utf8().ptr());
+}
+
+void _gdnative_report_loading_error(const godot_object *p_library, const char *p_what) {
+	String message = "Error loading GDNative file ";
+	GDNativeLibrary *library = (GDNativeLibrary *)p_library;
+
+	message += library->get_current_library_path() + ": reason \"" + p_what + "\"";
```
Comment (endragor) [inline modules/gdnative/gdnative/gdnative.cpp]: I think "reason" can be removed here so that there is more freedom in how the error is worded.

**Comment by endragor** [comment]
Yes, I think it's sufficient, thank you.

File: modules/gdnative/gdnative/gdnative.cpp (hunk 1 above)
Comment (karroffel) [reply modules/gdnative/gdnative/gdnative.cpp]: Right, will change it in as soon as I'm at my PC

**Pull request merged**